src/context.ts

```diff
@@ -0,0 +1,109 @@
+import { grpc } from '@improbable-eng/grpc-web'
+import { ThreadID } from '@textile/threads-id'
+import { Config } from '@textile/threads-client'
+
+type HostString = 'https://hub.textile.io:443' | 'https://hub.staging.textile.io:443' | 'http://127.0.0.1:3007' | string
+export const defaultHost: HostString = 'https://hub.textile.io:443'
+
+export interface ContextKeys {
+  /**
+   * Thread name. Specifies a mapping between human-readable name and a Thread ID.
+   */
+  ['x-textile-thread-name']?: string
+  /**
+   * Thread ID as a string. Should be generated with `ThreadID.toString()` method.
+   */
+  ['x-textile-thread']?: string
+  /**
+   * Session key. Used for various session contexts.
+   */
+  ['x-textile-session']?: string
+
+  /**
+   * Org slug/name. Used for various org session operations.
+   */
+  ['x-textile-org']?: string
+}
+
+/**
+ * Context provides immutable context management for gRPC credentials and config settings.
+ */
+export class Context implements Config {
```

Comment:
I love the go context pattern, so we've got something similar here. I think it'll be useful for a) specifying headers without having to know much about them, and b) mapping docs between js and go. Also, what a great design pattern! This version is quite opinionated in terms of which types of keys it'll accept (you have to extend it each time). Maybe in the future, I'll create a generic version of this. For now, this works great. Note also that it compiles with our existing Config interface, so we can move that stuff over more slowly.

Reply:
thinking through this. since `Context` implements `Config`, doesn't threads-client already accept `Context` (since it accepts `Config`)?  I guess you meant above that we need to include a few more of these headers. 

One rub here... I can't import Context in threads-client (since it'll be circular dep). So perhaps we split Context out of here?

Reply:
actually... it looks like i don't need to implement anything new from here since you are using the wrap methods primarily. so it _should_ just be gtg right?

Reply:
Correct

Reply:
The only sticking point is the 'hack' in the comment below. Since js-threads-client checks that the input isinstance of Config (and Context isn't) it basically 'wipes' out all the internal headers. Hence the 'hack' to patch it back on below. So if we adjust that check, we'll be just fine.